src/security/Handshake.cc

```diff
@@ -273,14 +273,16 @@ Security::HandshakeParser::parseModernRecord()
     // RFC 5246: MUST NOT send zero-length [non-application] fragments
     Must(record.fragment.length() || record.type == ContentType::ctApplicationData);
 
+    const auto expectMoreRecordBytes = !(tkRecords.atEnd() && !tkRecords.expectMore());
     if (currentContentType != record.type) {
         tkMessages.expectMore(false);
         parseMessages();
         Must(tkMessages.atEnd()); // no currentContentType leftovers
-        tkMessages.reset(record.fragment, true);
+        tkMessages.reset(record.fragment, expectMoreRecordBytes);
         currentContentType = record.type;
     } else {
-        tkMessages.reset(tkMessages.leftovers().append(record.fragment), true);
+        tkMessages.expectMore(expectMoreRecordBytes);
+        tkMessages.append(record.fragment);
```

Comment:
Done.

> that primary part asserts tkMessages.atEnd()

I have one concern: atEnd() does not check that everything has been committed (so that a subsequent rollback() would not mess things up). Though we should have called commit() by this time in message parsers - some new code may miss it.